[4.0] Redo the image editing api

dgrammatiko opened the pull request:
Pull Request is the second part for  the Tabs reworked code: https://github.com/joomla/joomla-cms/pull/34813

### Summary of Changes
Some changes to the API:
- Now it's async (used to be synchronous, but that won't work for many cases, current libraries, etc)
- Fixing all the problems with:
- - Empty fields on load
- - Save needed a click away to update the image
- - other inconsistencies


### Testing Instructions
#### This PR is not testable with the downladable version

Apply this PR!!!
You need to run `npm install`

Try to edit some images and using all the existing options

### Actual result BEFORE applying this Pull Request

Many bugs

### Expected result AFTER applying this Pull Request

No (obvious) bugs

### Documentation Changes Required

The old registration was something like:
```js
  // Register the Events
  Joomla.MediaManager.Edit.crop = {
    Activate(mediaData) {
      // Initialize
      initCrop(mediaData);
    },
    Deactivate() {
      if (!Joomla.MediaManager.Edit.crop.cropper) {
        return;
      }
      // Destroy the instance
      Joomla.MediaManager.Edit.crop.cropper.destroy();
    },
  };
```

Which is changed to:

```js
// Register the Events
window.addEventListener('media-manager-edit-init', () => {
  Joomla.MediaManager.Edit.plugins.crop = {
    Activate(image) {
      return new Promise((resolve /* , reject */) => {
        init(image);
        Joomla.MediaManager.Edit.plugins.crop.cropper
          .setAspectRatio(formElements.cropAspectRatioOption.value);
        resolve();
      });
    },
    Deactivate(image) {
      return new Promise((resolve /* , reject */) => {
        if (image.cropper) {
          image.cropper.destroy();
          delete Joomla.MediaManager.Edit.plugins.crop.cropper;
        }
        resolve();
      });
    },
  };
}, { once: true });

```

So most of the changes are just changes in the initialization step:
- there's an event for the registration, on the window, named 'media-manager-edit-init'
- The plugin name is now one level deeper, from `Joomla.MediaManager.Edit.crop` to `Joomla.MediaManager.Edit.plugins.crop`
- Both the `Activate` and `Deactivate` functions should return a promise.

@wilsonge sorry...

dgrammatiko commented:
This PR should be fixing:
https://github.com/joomla/joomla-cms/issues/33843
https://github.com/joomla/joomla-cms/issues/31859
https://github.com/joomla/joomla-cms/pull/33192

dgrammatiko committed:
wrong place for the max-width

Quy commented:
I have tested this item :white_check_mark: successfully on dc65bdbb72158788e34b5a1f367c11d7a7941b39<br /><br />THANK YOU!!!!<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/34885">issues.joomla.org/tracker/joomla-cms/34885</a>.</sub>

Quy commented:
Edit an image:
> Uncaught TypeError: this.Undo is undefined edit-images.js:99:5

Click Close button:
> Uncaught TypeError: newForm.task is undefined core.js:389:7

dgrammatiko committed:
Fixes

dgrammatiko commented:
> Uncaught TypeError: this.Undo is undefined edit-images.js:99:5

Fixing the lint issues without checking the code is not a great idea, should be ok now

dgrammatiko committed:
CS

ceford commented:
I have tested this item :white_check_mark: successfully on 627b449ec236b38486bb2e0310260e2d788d547e<br /><br />It all seems to work. I did notice some anomalies:
The aspect ratio box default was 'None'. The Edit screen does not have a Help button (it could use the media screen Help). Changing X-axis and Width did not show in the grid. The image is barely visible in 768x1024 screen mode and the grid shrinks strangely. Cannot edit (etc) from list view. In edit mode repeated clicking of Save moves the images left a little with each click (but the saved image is OK).<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/34885">issues.joomla.org/tracker/joomla-cms/34885</a>.</sub>

dgrammatiko commented:
> The aspect ratio box default was 'None'. 
Oops I change that in the other PR, will undo the changes here

> The Edit screen does not have a Help button (it could use the media screen Help). 

This is out of the scope of this PR but should be easy to add a help button here: https://github.com/joomla/joomla-cms/blob/193408176112c05055d9e005c0b82070cf0c5aef/administrator/components/com_media/src/View/File/HtmlView.php#L75

> Changing X-axis and Width did not show in the grid.
> The image is barely visible in 768x1024 screen mode and the grid shrinks strangely.
> In edit mode repeated clicking of Save moves the images left a little with each click (but the saved image is OK).
I guess this is for crop, investigating all three issues


> Cannot edit (etc) from list view.
This is something for the Media manager

Quy commented:
For crop, the image expands height/width accordingly, but not under the other actions. Is this intentional?

dgrammatiko commented:
> For crop, the image expands height/width accordingly, but not under the other actions. Is this intentional?

The image has a `max-width: 100%` so it should adapt to the parent's size. I didn't touch that part (or the logic), just updated the initialization logic and fixed some obvious mistakes (re attaching listeners, etc). Anyways I will check this again

dgrammatiko commented:
> For crop, the image expands height/width accordingly, but not under the other actions. Is this intentional?

Should be fine now

Quy commented:
It is no longer side by side. The settings are at the top and image at the bottom.

dgrammatiko commented:
> It is no longer side by side. The settings are at the top and image at the bottom.

Oops, sorry, handling too many PRs is not easy

Quy commented:
The image still expands as seen between crop and resize.

![34885-crop](https://user-images.githubusercontent.com/368084/126933938-1738e25c-890f-4c1d-baaf-326b0b3bc687.jpg)

![34885-resize](https://user-images.githubusercontent.com/368084/126933944-3bd1e8a0-8cfc-431b-8cb9-6576121c64a1.jpg)

dgrammatiko committed:
Update edit-images.es6.js

dgrammatiko committed:
Update crop.es6.js

dgrammatiko commented:
Image size corrected

dgrammatiko committed:
tests test

richard67 commented:
@dgrammatiko It needs to solve conflicts and update the branch here after recent merge into 4.0-dev.

dgrammatiko commented:
> @dgrammatiko It needs to solve conflicts and update the branch here after recent merge into 4.0-dev.

Yup, and now I hope tests are green

richard67 commented:
@dgrammatiko You might have to update your branch again because just right now George merged the changed tests into 4.0-dev.

dgrammatiko committed:
Redo

richard67 commented:
@dgrammatiko System tests failed so it seems you have updated your branch just 2 minutes too early.

dgrammatiko committed:
Styling, Crop previous settings

brianteeman commented on build/media_source/com_media/scss/components/_media-edit.scss:
```diff
@@ -8,7 +8,7 @@
 
   joomla-tab-element {
-    grid-template-columns: repeat(4, minmax(240px, 1fr));
+    grid-template-columns: repeat(auto-fit, minmax(240px,1fr));
```
```suggestion
    grid-template-columns: repeat(auto-fit, minmax(240px, 1fr));
```

brianteeman commented on build/media_source/plg_media-action_rotate/js/rotate.es6.js:
```diff
@@ -2,36 +2,31 @@
-Joomla.MediaManager.Edit = Joomla.MediaManager.Edit || {};
+// Update image
+const rotate = (angle, image) => {
+  // The canvas where we will resize the image
```
```suggestion
  // The canvas where we will rotate the image
```

dgrammatiko committed:
Update build/media_source/com_media/scss/components/_media-edit.scss

Co-authored-by: Brian Teeman <brian@teeman.net>

dgrammatiko committed:
Update build/media_source/plg_media-action_rotate/js/rotate.es6.js

Co-authored-by: Brian Teeman <brian@teeman.net>

richard67 commented:
All CI checks green ... there must be something wrong ;-)

wilsonge commented:
I don't think the tests cover much of the image editing :P Mainly the list view :D

dgrammatiko committed:
Fixes

dgrammatiko commented:
After https://github.com/joomla/joomla-cms/pull/34885/commits/1c8e84be00e194d61ceaea369b3a85ae3726557f I think all should be fine

dgrammatiko committed:
obviously not

richard67 commented:
@dgrammatiko Javascript CS: https://ci.joomla.org/joomla/joomla-cms/46162/1/21

dgrammatiko committed:
Rotate fixes

dgrammatiko commented:
This should be ready for testing now

dgrammatiko committed:
Comment again

dgrammatiko commented:
> I don't think the tests cover much of the image editing :P Mainly the list view :D

The best tests, navigate to the page read the title. Test is passing. Yeah...

wilsonge commented:
> The best tests, navigate to the page read the title. Test is passing. Yeah...

Lots of the tests still fail on bad changes though :P What does that say :D

dgrammatiko commented:
> What does that say :D

Unstable server?

wilsonge commented:
So the crop filters don't apply - but also don't error if you pick e.g. a portrait crop for a landscape image. I also got a 403 back when trying to save it. It's a put request to `http://localhost/~george/joomla-cms/administrator/index.php?option=com_media&format=json&mediatypes=0,1,2,3&task=api.files&path=local-images:/banner-og-github.png` - will test properly later just in case it's something locally. But somehow doubt it as it's just a simple file upload then edit.

dgrammatiko commented:
@wilsonge make sure you have cleared the browser cache, but this is also a bug

dgrammatiko committed:
Cache invalidation

Quy commented:
Edit an image.
Click Resize tab.
Click Save button.
Image saved.
Click Save button.
Error in console: 
> Illegal mime type detected: application/octet-stream

dgrammatiko committed:
Nulify this.xhr

dgrammatiko committed:
Fix the upload loagic

- Fix the XSS with the urls

dgrammatiko committed:
CS

wilsonge committed:
Typo

wilsonge commented:
<img width="1274" alt="Screenshot 2021-07-27 at 18 21 14" src="https://user-images.githubusercontent.com/1986000/127199322-7341c3ab-54e3-4e61-8d8b-10d16966c542.png">

Hit the 90 degree rotate button twice. You get kinda weird behaviour.

dgrammatiko committed:
Fix rotate rendering

dgrammatiko commented:
> Hit the 90 degree rotate button twice. You get kinda weird behaviour.

Fixed

dgrammatiko committed:
grrrr

wilsonge commented:
If you manually enter text for the crop it seems to have no affect. If you adjust the crop box on the image it works fine. 

The clicking of 90 degree's rotation is better but not fixed too.

dgrammatiko commented:
> The clicking of 90 degree's rotation is better but not fixed too.

What's the problem now? The image should cover all the right area (assuming it's big enough)

> If you manually enter text for the crop it seems to have no effect. If you adjust the crop box on the image it works fine.

@Quy also mentioned this but the initialization code for cropper is the same AFAICT

wilsonge commented:
> What's the problem now? The image should cover all the right area (assuming it's big enough)

Same issue. I just have to click 90 degree rotation 4 times instead of twice :) I'm not speed clicking either to try and break things. Sometimes if you wait long enough the screen area resizes itself back to the correct size.

wilsonge commented:
I also think the aspect ratio dropdown is broken?

dgrammatiko committed:
Keep the same einstance of cropper

dgrammatiko commented:
@wilsonge  Crop should be fine again with https://github.com/joomla/joomla-cms/pull/34885/commits/910ac38d0f59f0ba92bcb5ce65689fff75739879

dgrammatiko commented:
> Same issue. I just have to click 90 degree rotation 4 times instead of twice :) I'm not speed clicking either to try and break things. Sometimes if you wait long enough the screen area resizes itself back to the correct size.

Ok, so this was a 2 part solution:
- there was some weird css on the parent element of the image
- browser needed some time to rethink the layout/painting

Give it some hard reloading before testing

dgrammatiko committed:
nooo

wilsonge commented:
OK So cropper works great now! Way better. But now from that commit rotation doesn't work. It never leaves the cropper view

dgrammatiko committed:
again

wilsonge commented:
<img width="1273" alt="Screenshot 2021-07-27 at 20 34 08" src="https://user-images.githubusercontent.com/1986000/127216541-16977abe-b02e-4d77-a73c-24af6d0606e6.png">

It's all broken 😆

wilsonge commented on build/media_source/com_media/js/edit-images.es6.js:
```diff
@@ -2,304 +2,372 @@
+
+  async activate(name) {
+    this.imagePreview = null;
+    this.createImageContainer(this.current.contents);
```
```suggestion
    this.createImageContainer(this.current);
```

So this fixes the initialisation error. But then i get 4 copies of the image loaded at full screen. So it's not really a good fix :)

dgrammatiko committed:
again

wilsonge committed:
JS CS

wilsonge commented:
![Screenshot 2021-07-27 at 22 29 55](https://user-images.githubusercontent.com/1986000/127230048-21c061bc-82c2-40c2-8dc8-f7ee01b20397.png)

I mean rotate may well be acting the same as before. But not sure this is intended :D However it's way better than it has been all evening.

wilsonge commented:
Thanks!

dgrammatiko commented:
Release the Kraken

![source](https://user-images.githubusercontent.com/3889375/127231124-b67ffa1c-54e9-492c-956d-35b5b7d2a348.gif)

richard67 commented:
@dgrammatiko There is still much to be fixed.

dgrammatiko commented:
@richard67 I know, but a new RC is always a good thing... 😉

laoneo commented:
Sorry guys, but since this pr the reset button is not working anymore.

richard67 commented:
Has it ever worked? ;-)

richard67 commented:
Sorry, was joke. Could you open an issue?
